number/number.go

```
@@ -0,0 +1,156 @@
+func Search(c *messagebird.Client, cc string, listParams *NumberListParams) (*NumberSearchingList, error) {
+	uri := GetPath(listParams, pathNumbersAvailable+"/"+cc)
+
+	numberList := &NumberSearchingList{}
+	if err := request(c, numberList, http.MethodGet, uri, nil); err != nil {
+		return nil, err
+	}
+
+	return numberList, nil
+}
+
+// Read get a purchased phone number
+func Read(c *messagebird.Client, phoneNumber string) (*Number, error) {
+	if len(phoneNumber) < 5 {
+		return nil, fmt.Errorf("a phoneNumber is too short")
+	}
+
+	uri := fmt.Sprintf("%s/%s", pathNumbers, phoneNumber)
+
+	number := &Number{}
+	if err := request(c, number, http.MethodGet, uri, nil); err != nil {
+		return nil, err
+	}
+
+	return number, nil
+}
+
+// GetPath get the full path for the request
+func GetPath(listParams *NumberListParams, path string) string {
+	params := paramsForMessageList(listParams)
+	return fmt.Sprintf("%s?%s", path, params.Encode())
+}
+
+// paramsForMessageList build query params
+func paramsForMessageList(params *NumberListParams) *url.Values {
+	urlParams := &url.Values{}
+
+	if params == nil {
+		return urlParams
+	}
+
+	if len(params.Features) > 0 {
+		paramsForArrays("features", "^(sms|voice|mms)$", params.Features, urlParams)
+	}
+
+	if len(params.Type) > 0 {
+		paramsForArrays("type", "^(mobile|mobile|premium_rate)$", params.Type, urlParams)
+	}
+
+	if params.Number != "" {
+		urlParams.Set("number", params.Number)
+	}
+	if params.Country != "" {
+		urlParams.Set("country", params.Country)
+	}
+	if params.Limit != 0 {
+		urlParams.Set("limit", strconv.Itoa(params.Limit))
+	}
+
+	if params.SearchPattern != "" {
+		urlParams.Set("search_pattern", params.SearchPattern)
+	}
+
+	if params.Offset != 0 {
+		urlParams.Set("offset", strconv.Itoa(params.Offset))
+	}
+
+	return urlParams
+}
+
+// paramsForArrays build query for array params
+func paramsForArrays(field string, pattern string, array []string, urlParams *url.Values) {
+	r, _ := regexp.Compile(pattern)
```

Compiling regexp is not that cheap, so it may be worthwhile to do something like `var someRe = regexp.MustCompile(regexp.Compile("pattern"))` at the package level and re-use that.

---

Also, it makes sense to bubble up an error when one of the values is considered invalid (rather than silently ignoring it).